### PR TITLE
Javaparser

### PR DESCRIPTION
**Pull request by Bohdankm22**

### CODE REVIEW

**Comment by mernst** [comment]
You should leave the `master` branch of your fork unchanged.
A better way to formulate the pull request would be from your fork's branch into upstream's master.

**Comment by Bohdankm22** [comment]
@mernst 
The first my pull request was to my the fork master branch. 
This is the second pull request to there. After merging into the master of my fork all my changes will be in there.
Why should this be to upstream master?

**Comment by mernst** [comment]
I see.  I will review this.

In the future, it's best to leave the master branch of a fork clean, to accomodate pulling from upstream, and do all your work on branches.
This is mentioned at https://homes.cs.washington.edu/~mernst/advice/github-pull-request.html .

File: framework/src/org/checkerframework/framework/stub/StubParser.java [hunk 1]
```diff
@@ -729,28 +712,37 @@ private void annotate(AnnotatedTypeMirror atype, Type typeDef) {
...
+                            declType.getTypeArguments().get().get(i));
                 }
             }
         } else if (atype.getKind() == TypeKind.WILDCARD) {
             AnnotatedWildcardType wildcardType = (AnnotatedWildcardType) atype;
             WildcardType wildcardDef = (WildcardType) typeDef;
-            if (wildcardDef.getExtends() != null) {
-                annotate(wildcardType.getExtendsBound(), wildcardDef.getExtends());
+            if (wildcardDef.getExtendedType() != null) {
+                // TODO Simplify the logic
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubParser.java]: Handle TODO??

File: framework/src/org/checkerframework/framework/stub/StubParser.java [hunk 2]
```diff
@@ -729,28 +712,37 @@ private void annotate(AnnotatedTypeMirror atype, Type typeDef) {
...
+                if (wildcardDef.getExtendedType().isPresent()) {
+                    annotate(wildcardType.getExtendsBound(), wildcardDef.getExtendedType().get());
+                }
                 annotate(wildcardType.getSuperBound(), typeDef.getAnnotations());
-            } else if (wildcardDef.getSuper() != null) {
-                annotate(wildcardType.getSuperBound(), wildcardDef.getSuper());
+            } else if (wildcardDef.getSuperType() != null) {
+                // TODO Simplify the logic
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubParser.java]: Handle TODO??

File: framework/src/org/checkerframework/framework/stub/StubParser.java [hunk 3]
```diff
@@ -786,17 +778,20 @@ private void parseConstructor(
...
         }
 
-        if (methodType.getReceiverType() == null
-                && decl.getReceiverAnnotations() != null
-                && !decl.getReceiverAnnotations().isEmpty()) {
+        if (methodType.getReceiverType() == null) {
+            // DO nothing
+            // TODO remove this block after resolving TODO below
+        } else if (methodType.getReceiverType() == null
+                // TODO Fix - here should be receiver annotations. Remove && false
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubParser.java]: Fix this.

File: framework/src/org/checkerframework/framework/stub/StubParser.java [hunk 4]
```diff
@@ -884,15 +884,19 @@ private void annotateTypeParameters(
...
+        assert (typeElt.getSimpleName().contentEquals(typeDecl.getNameAsString())
+                        || typeDecl.getNameAsString()
+                                .endsWith("$" + typeElt.getSimpleName().toString()))
                 : String.format("%s  %s", typeElt.getSimpleName(), typeDecl.getName());
 
-        Map<Element, BodyDeclaration> result = new HashMap<>();
+        Map<Element, BodyDeclaration<?>> result = new HashMap<>();
 
-        for (BodyDeclaration member : typeDecl.getMembers()) {
+        // TODO Fix the problem with raw type
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubParser.java]: What is the problem that must be fixed?

File: framework/src/org/checkerframework/framework/stub/StubParser.java [hunk 5]
```diff
@@ -918,17 +922,14 @@ private void annotateTypeParameters(
...
-                                            ciDecl.getBeginLine(),
-                                            typeDecl.getName(),
-                                            ciDecl.getName())
-                                    + "\n"
-                                    + String.format(
-                                            "  Instead, write the nested class as a top-level class:%n    class %s { ... }%n    class %s$%s { ... }%n",
-                                            typeDecl.getName(),
-                                            typeDecl.getName(),
-                                            ciDecl.getName()));
+                                    "Warning: ignoring nested class in %s at line %d; "
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubParser.java]: Does the stub parser still ignore nested classes?  I don't think it should do so.  Please fix the behavior and add a test for nested classes.

File: framework/src/org/checkerframework/framework/stub/StubParser.java [hunk 6]
```diff
@@ -956,21 +957,22 @@ private AnnotatedDeclaredType findType(
...
         }
         return null;
     }
 
     private ExecutableElement findElement(TypeElement typeElt, MethodDeclaration methodDecl) {
-        final String wantedMethodName = methodDecl.getName();
+        final String wantedMethodName = methodDecl.getNameAsString();
         final int wantedMethodParams =
                 (methodDecl.getParameters() == null) ? 0 : methodDecl.getParameters().size();
+        // TODO Make sure assignment was changed right
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubParser.java]: Check, then remove comment.

File: framework/src/org/checkerframework/framework/stub/StubUtil.java [hunk 7]
```diff
@@ -68,20 +59,21 @@
...
-        TypeDeclaration type =
+            VariableElement field, StubUnit indexFile) {
+        TypeDeclaration<?> type =
                 findDeclaration((TypeElement) field.getEnclosingElement(), indexFile);
         if (type == null) {
             return null;
         }
 
-        for (BodyDeclaration member : type.getMembers()) {
+        NodeList<BodyDeclaration<?>> members = type.getMembers();
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubUtil.java]: Why is this change necessary?

File: framework/src/org/checkerframework/framework/stub/StubUtil.java [hunk 8]
```diff
@@ -95,17 +87,18 @@
...
+        TypeDeclaration<?> type =
                 findDeclaration((TypeElement) method.getEnclosingElement(), indexFile);
         if (type == null) {
             return null;
         }
 
         String methodRep = toString(method);
 
-        for (BodyDeclaration member : type.getMembers()) {
+        NodeList<BodyDeclaration<?>> members = type.getMembers();
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubUtil.java]: Why is this change necessary?

File: framework/src/org/checkerframework/framework/stub/StubUtil.java [hunk 9]
```diff
@@ -269,10 +262,11 @@ public void visit(MethodDeclaration n, Void arg) {
 
         @Override
         public void visit(Parameter n, Void arg) {
-            if (n.getId().getArrayCount() > 0) {
+            // TODO Implement this check
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubUtil.java]: Handle all todos before merging pull request.

File: framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java [hunk 10]
```diff
@@ -182,23 +173,26 @@ private static void convert(AScene scene, InputStream in, OutputStream out)
...
                 List<ImportDeclaration> impDecls = cu.getImports();
-                PackageDeclaration pkgDecl = cu.getPackage();
-                for (TypeDeclaration typeDecl : typeDecls) {
-                    ToIndexFileConverter converter =
-                            new ToIndexFileConverter(pkgDecl, impDecls, scene);
-                    String pkgName = converter.pkgName;
-                    String name = typeDecl.getName();
-                    if (!pkgName.isEmpty()) {
-                        name = pkgName + "." + name;
+                // TODO Refactor. Replace with simplier check
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java]: TODO comment.

File: framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java [hunk 11]
```diff
@@ -313,8 +307,8 @@ public Void visit(MethodDeclaration decl, AElement elem) {
         List<Parameter> params = decl.getParameters();
         List<TypeParameter> typeParams = decl.getTypeParameters();
         List<AnnotationExpr> rcvrAnnos = decl.getReceiverAnnotations();
-        BlockStmt body = decl.getBody();
-        StringBuilder sb = new StringBuilder(decl.getName()).append('(');
+        BlockStmt body = decl.getBody().isPresent() ? decl.getBody().get() : null;
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java]: `orElse(null)` is a better idiom to use than `isPresent() ? ... : null`.

File: framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java [hunk 12]
```diff
@@ -359,11 +353,15 @@ public Void visit(MethodDeclaration decl, AElement elem) {
     @Override
     public Void visit(ObjectCreationExpr expr, AElement elem) {
         ClassOrInterfaceType type = expr.getType();
-        AClass clazz = scene.classes.vivify(type.getName());
-        Expression scope = expr.getScope();
-        List<Type> typeArgs = expr.getTypeArgs();
-        List<Expression> args = expr.getArgs();
-        List<BodyDeclaration> decls = expr.getAnonymousClassBody();
+        AClass clazz = scene.classes.vivify(type.getNameAsString());
+        // TODO Make sure object is present
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java]: TODO comments.  Probably use `orElse`?

File: framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java [hunk 13]
```diff
@@ -496,7 +499,7 @@ private void visitInnerType(Type type, InnerTypeLocation loc) {
                     /**
                      * Extends type path by one element.
                      *
-                     * @see TypePathEntry.fromBinary
+                     * <p>//* @see TypePathEntry.fromBinary
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java]: Undo this change.

File: framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java [hunk 14]
```diff
@@ -624,8 +629,9 @@ private String resolve(String className) {
     /**
      * Combines an import with a name, yielding a fully-qualified name.
      *
-     * @param prefix name of imported package
-     * @param base name of class, possibly qualified
+     * <p>TODO Fix the problem with Java docs //* @param prefix name of imported package //* @param
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java]: The TODO comment has broken the formatting.  Anyway, address it.

File: framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java [hunk 15]
```diff
@@ -56,6 +56,7 @@ public static TestConfigurationBuilder getDefaultConfigurationBuilder(
                         .addOption("-Xlint:unchecked")
                         .addOption("-XDrawDiagnostics") // use short javac diagnostics
                         .addOption("-AprintErrorStack")
+                        // .addOption("-AstubDebug")
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java]: Don't put temporary comments in a pull request.

File: checker/src/org/checkerframework/checker/guieffect/jdk.astub [hunk 16]
```diff
@@ -1,20 +1,24 @@
 package java.lang;
 import org.checkerframework.checker.guieffect.qual.*;
 @PolyUIType class Object {
-    @SafeEffect Class<? extends Object> getClass() @PolyUI;
+    @SafeEffect Class<? extends Object> getClass(@PolyUI Object this) ;
+}
+
+@PolyUIType interface Runnable {
```
Comment (Bohdankm22) [inline checker/src/org/checkerframework/checker/guieffect/jdk.astub]: The interface was not removed. It was transferred to the another park of the stub file because the java.lang package was already defined and according to the stub file format each package declaration may appear only once per stub file.

File: checker/src/org/checkerframework/checker/guieffect/jdk.astub [hunk 17]
```diff
@@ -112,12 +116,5 @@ public abstract class JTextComponent extends JComponent implements Scrollable, A
     @SafeEffect public void setText(String t);
 }
 
-
-package java.lang;
-
-@PolyUIType interface Runnable {
```
Comment (Bohdankm22) [inline checker/src/org/checkerframework/checker/guieffect/jdk.astub]: The interface was not removed. It was transferred to the another park of the stub file because the java.lang package was already defined and according to the stub file format each package declaration may appear only once per stub file.

File: checker/jtreg/nullness/issue824/Second.java [hunk 18]
```diff
@@ -14,7 +14,8 @@ public static void one(
         First.method(supplier, callable);
     }
 
-    public static void two(First.Supplier<Integer> supplier, First.Callable<Object> callable) {
+    public static void two(
+            First.Supplier<Integer> supplier, First.Callable<@Nullable Object> callable) {
```
Comment (Bohdankm22) [inline checker/jtreg/nullness/issue824/Second.java]: I am not sure, but it was not working without the change.

File: framework/src/org/checkerframework/framework/stub/StubParser.java [hunk 19]
```diff
@@ -882,68 +882,56 @@ private void annotateTypeParameters(
...
-                                "Ignoring element of type %s in getMembers", member.getClass()));
             }
+        } else if (member instanceof ClassOrInterfaceDeclaration) {
+            Element elt = findElement(typeElt, (ClassOrInterfaceDeclaration) member);
+            if (elt != null) {
+                putNew(result, elt, member);
+            }
+        } else {
+            stubWarnIfNotFound(
+                    String.format("Ignoring element of type %s in getMembers", member.getClass()));
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubParser.java]: `getMembers` seems incorrect in this comment.
(It's not your fault, but could you fix it?)

File: framework/src/org/checkerframework/framework/stub/StubParser.java [hunk 20]
```diff
@@ -956,21 +944,41 @@ private AnnotatedDeclaredType findType(
...
         if (debugStubParser) {
             for (AnnotatedDeclaredType superType : types) {
-                stubDebug(String.format("  %s%n", superType));
+                stubDebug(String.format("  %s", superType));
+            }
+        }
+        return null;
+    }
+
+    private Element findElement(TypeElement typeElt, ClassOrInterfaceDeclaration ciDecl) {
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubParser.java]: Could you add Javadoc for each method you wrote?  (I realize that not all the existing methods have documentation, but yours can be better.)

File: framework/src/org/checkerframework/framework/stub/StubParser.java [hunk 21]
```diff
@@ -1060,20 +1069,24 @@ private PackageElement findPackage(String packageName) {
...
-                            + " => "
-                            + m.get(key)
-                            + LINE_SEPARATOR
-                            + "while adding: "
-                            + LINE_SEPARATOR
-                            + "  "
-                            + key
-                            + " => "
-                            + value);
+            // TODO Check does it affect anything
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubParser.java]: Please address this todo comment and reinstate the code.

File: framework/src/org/checkerframework/framework/stub/StubUtil.java [hunk 22]
```diff
@@ -269,11 +259,6 @@ public void visit(MethodDeclaration n, Void arg) {
 
         @Override
         public void visit(Parameter n, Void arg) {
-            if (n.getId().getArrayCount() > 0) {
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/StubUtil.java]: Double-check this removal:  can the code be retained?
The goal is to prevent 
  `int a[]`
from appearing in stub files, and to make users write
  `int[] a`
instead.

File: framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java [hunk 23]
```diff
@@ -442,7 +433,8 @@ private static Void visitInnerTypes(Type type, final ATypeElement elem) {
                 new GenericVisitorAdapter<Void, InnerTypeLocation>() {
                     @Override
                     public Void visit(ClassOrInterfaceType type, InnerTypeLocation loc) {
-                        List<Type> typeArgs = type.getTypeArgs();
+                        // TODO Make sure object is present
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java]: Address todo, then remove the comment.

File: framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java [hunk 24]
```diff
@@ -451,13 +443,15 @@ public Void visit(ClassOrInterfaceType type, InnerTypeLocation loc) {
                         return null;
                     }
 
-                    @Override
+                    // TODO It was a overriden method.
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java]: What is to be done for this TODO comment?

File: framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java [hunk 25]
```diff
@@ -451,13 +443,15 @@ public Void visit(ClassOrInterfaceType type, InnerTypeLocation loc) {
...
+                    // @Override
                     public Void visit(ReferenceType type, InnerTypeLocation loc) {
                         InnerTypeLocation ext = loc;
-                        int n = type.getArrayCount();
+                        int n = type.getArrayLevel();
                         for (int i = 0; i < n; i++) {
                             ext = extendedTypePath(ext, 1, 0);
-                            for (AnnotationExpr expr : type.getAnnotationsAtLevel(i)) {
+                            // TODO There was a getting annotations at level i, but it was changed for getting Annotation not depend on level
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java]: This seems wrong to me.  Have you tested with multi-dimensional arrays and annotations on different levels of them?

File: framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java [hunk 26]
```diff
@@ -533,33 +527,34 @@ public String visit(ClassOrInterfaceType type, Void v) {
...
+                            case SHORT:
                                 return "S";
                             default:
                                 throw new IllegalArgumentException(
                                         "baseTypeName(): unknown primitive type " + type);
                         }
                     }
 
-                    @Override
+                    // TODO It was a overriden method.
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java]: Please address this TODO

File: framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java [hunk 27]
```diff
@@ -624,8 +619,6 @@ private String resolve(String className) {
     /**
      * Combines an import with a name, yielding a fully-qualified name.
      *
-     * @param prefix name of imported package
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java]: Rather than removing these, update them to reflect the current parameter names (and semantics, if that has changed).

Comment (Bohdankm22) [inline framework/src/org/checkerframework/framework/stub/ToIndexFileConverter.java]: Addressed in https://github.com/Bohdankm22/checker-framework/pull/2/commits/d7f9c018fd0a1ff5fd4cafbfbba1197c00cfdafa

File: framework/src/org/checkerframework/framework/stub/StubUtil.java (hunk 22 above)
Comment (Bohdankm22) [inline framework/src/org/checkerframework/framework/stub/StubUtil.java]: Addressed in https://github.com/Bohdankm22/checker-framework/pull/2/commits/5000ebbee9687b28554a3adeff8e14a894ef92c0

Comment (Bohdankm22) [inline framework/src/org/checkerframework/framework/stub/StubUtil.java]: I debugged parsing arrays in the javaparser. If array brackets are written after variable name 
`int a[]`
it moves it in type and behave the same way as if it was written like
`int[] a`.
Another example:
`int[] a[]` it treated by the javaparser like `int[][] a`.

Hence this check is not needed.

**Comment by Bohdankm22** [comment]
All requested changes were addressed. Merging the pull request.
